lex.go

```diff
@@ -63,15 +107,35 @@ func tokenize(reader io.Reader, tokenCh chan NgxToken) {
 	depth := 0
 	var la, quote string
 
+	// check if the lua token is directive, if token line is already processd, then this is not the directive
+	nextTokenIsDirective := make(map[int]tokenInfo)
+
 	scanner := bufio.NewScanner(reader)
 	scanner.Split(bufio.ScanRunes)
 
 	emit := func(line int, quoted bool, err error) {
 		tokenCh <- NgxToken{Value: token.String(), Line: line, IsQuoted: quoted, Error: err}
+		nextTokenIsDirective[line] = tokenInfo{LineNumberExists: true}
 		token.Reset()
 		lexState = skipSpace
 	}
 
+	var externalLexers map[string]ExtLexer
+	var externalScanner *extScanner
+	for _, ext := range options.ExternalLexers {
+		if externalLexers == nil {
+			externalLexers = make(map[string]ExtLexer)
+		}
+
+		if externalScanner == nil {
+			externalScanner = &extScanner{scanner: scanner, tokenLine: tokenLine}
+		}
+
+		for _, d := range ext.Register(externalScanner) {
+			externalLexers[d] = ext
```

Comment:
I added code to emit an error token and exit

Reply:
I thought we talked offline about allowing it. The use case for that I think would be to provide an alternate implementation that "overrides" another extension in case you wanted it to behave differently for a subset of tokens?

I don't have a specific use case in mind though, so maybe this version is OK if other reviews prefer this strategy.